Mathlib/LinearAlgebra/Matrix/Swap.lean

```diff
@@ -0,0 +1,143 @@
+`Mathlib.LinearAlgebra.Matrix.Transvection`.
+
+## Implementation detail
+
+This is a thin wrapper around `(Equiv.swap i j).toPEquiv.toMatrix`.
+-/
+
+namespace Matrix
+
+section Def
+
+variable {R n : Type*} [Zero R] [One R] [DecidableEq n]
+
+variable (R) in
+/-- The swap matrix `swap R i j` is the identity matrix with the
+`i`-th and `j`-th rows modified such that multiplying by it on the
+left (resp. right) corresponds to swapping the `i`-th and `j`-th row (resp. column). -/
+def swap (i j : n) : Matrix n n R :=
+  (Equiv.swap i j).permMatrix R
+
+lemma swap_comm (i j : n) :
+    swap R i j = swap R j i := by
+  simp only [swap, Equiv.swap_comm]
+
+@[simp]
+lemma transpose_swap (i j : n) : (swap R i j).transpose = swap R i j := by
+  simp [swap]
+
+@[simp]
+lemma conjTranspose_swap {R : Type*} [Semiring R] [StarRing R] (i j : n) :
+    (swap R i j).conjTranspose = swap R i j := by
+  simp [swap]
+
+end Def
+
+section
+
+variable {R n : Type*} [Semiring R] [DecidableEq n]
+
+@[simp]
+lemma map_swap {S : Type*} [Semiring S] (f : R →+* S) (i j : n) :
+    (swap R i j).map f = swap S i j := by
+  simp [swap]
+
+variable [Fintype n]
+
+lemma swap_mulVec (i j : n) (a : n → R) :
+    swap R i j *ᵥ a = a ∘ Equiv.swap i j := by
+  simp [swap, PEquiv.toMatrix_toPEquiv_mulVec]
+
+lemma vecMul_swap (i j : n) (a : n → R) :
+    a ᵥ* swap R i j = a ∘ Equiv.swap i j := by
+  simp [swap, PEquiv.vecMul_toMatrix_toPEquiv]
+
+@[simp]
+lemma swap_mulVec_apply (i j : n) (a : n → R) :
+    (swap R i j *ᵥ a) i = a j := by
+  simp [swap, PEquiv.toMatrix_toPEquiv_mulVec]
+
+@[simp]
+lemma vecMul_swap_apply (i j : n) (a : n → R) :
+    (a ᵥ* swap R i j) i = a j := by
+  simp [swap, PEquiv.vecMul_toMatrix_toPEquiv]
+
+/-- Multiplying with `swap R i j` on the left swaps the `i`-th row with the `j`-th row. -/
+@[simp]
+lemma swap_mul_apply_left (a i j : n) (g : Matrix n n R) :
+    (swap R i j * g) i a = g j a := by
+  simp [swap, PEquiv.toMatrix_toPEquiv_mul]
+
+/-- Multiplying with `swap R i j` on the left swaps the `j`-th row with the `i`-th row. -/
+@[simp]
+lemma swap_mul_apply_right (a i j : n) (g : Matrix n n R) :
+    (swap R i j * g) j a = g i a := by
+  rw [swap_comm, swap_mul_apply_left]
+
+lemma swap_mul_of_ne {a b i j : n} (hai : a ≠ i) (haj : a ≠ j) (g : Matrix n n R) :
+    (swap R i j * g) a b = g a b := by
+  simp [swap, PEquiv.toMatrix_toPEquiv_mul, Equiv.swap_apply_of_ne_of_ne hai haj]
+
+/-- Multiplying with `swap R i j` on the right swaps the `i`-th column with the `j`-th column. -/
+@[simp]
+lemma mul_swap_apply_left (a i j : n) (g : Matrix n n R) :
+    (g * swap R i j) a i = g a j := by
+  simp [swap, PEquiv.mul_toMatrix_toPEquiv]
+
+/-- Multiplying with `swap R i j` on the right swaps the `j`-th column with the `i`-th column. -/
+@[simp]
+lemma mul_swap_apply_right (a i j : n) (g : Matrix n n R) :
+    (g * swap R i j) a j = g a i := by
+  rw [swap_comm, mul_swap_apply_left]
```

Comment:
Shouldn't `g` be a rectangular matrix here (and perhaps in all the other lemmas too)?